### PR TITLE
[TwigComponent] Allow input props to have the same name as context variables

### PR DESCRIPTION
**Pull request by squrious**
| Q            | A   |
|--------------|-----|
| Bug fix?     | yes |
| New feature? | no  |
| Issues       | -   |
| License      | MIT |

It looks like #1652 introduced an issue with input props and parent context.

If:
- A component prop has the same name as a variable in the parent context
- The context variable is used to set the prop value as an input prop
- The component is rendered with the embed strategy

Then, the default value of the prop is used instead of the one passed to the component.

Example: 

```twig
{# templates/components/Hello.html.twig #}

{% props name = 'John' %}
<div {{ attributes }}>
    Hello {{ name }}!
</div>
```

### Render embedded

```twig
{% set name = 'Bryan' %}
<twig:Hello :name="name"></twig:Hello>
```

Wrong output: 
```html
<div>
    Hello John!
</div>
```

### Render with function

```twig
{% set name = 'Bryan' %}
<twig:Hello :name="name" />
```

Correct output: 
```html
<div>
    Hello Bryan!
</div>
```

### CODE REVIEW

File: src/TwigComponent/src/ComponentRenderer.php [hunk 1]
```diff
@@ -180,7 +182,7 @@ private function exposedVariables(object $component, bool $exposePublicProps): \
             $name = $attribute->name ?? (str_starts_with($method->name, 'get') ? lcfirst(substr($method->name, 3)) : $method->name);
 
             if ($method->getNumberOfRequiredParameters()) {
-                throw new \LogicException(sprintf('Cannot use %s on methods with required parameters (%s::%s).', ExposeInTemplate::class, $component::class, $method->name));
+                throw new \LogicException(sprintf('Cannot use "%s" on methods with required parameters (%s::%s).', ExposeInTemplate::class, $component::class, $method->name));
```
Comment (smnandre) [inline src/TwigComponent/src/ComponentRenderer.php:184]: ❤️

File: src/TwigComponent/src/ComponentRenderer.php [hunk 2]
```diff
@@ -113,12 +113,14 @@ private function preRender(MountedComponent $mounted, array $context = []): PreR
...
+        $propsKeys = array_keys($props);
         $variables = array_merge(
             // first so values can be overridden
             $context,
             // add the context in a separate variable to keep track
-            // of what is coming from outside the component
-            ['__context' => $context],
+            // of what is coming from outside the component, excluding props
+            // as they override initial context values
+            ['__context' => array_filter($context, static fn ($key) => !\in_array($key, $propsKeys), \ARRAY_FILTER_USE_KEY)],
```
Comment (smnandre) [inline src/TwigComponent/src/ComponentRenderer.php]:
```suggestion
            ['__context' => array_diff_key($context, $props)],
```

(I have a doubt)

**Comment by smnandre** [comment]
Maybe add a test case when name is passed with "null" ?

**Comment by squrious** [comment]
Thanks for the review @smnandre!

> Maybe add a test case when name is passed with "null" ?

I wasn't sure about that. If you pass null, the prop is considered unset, so yes we can't pass a null value explicitly. 

But there is a systematic use of `isset` in the PropsNode class, so I thought it was intended for some reason. Would you prefer to replace them with  `array_key_exists`?

File: src/TwigComponent/src/ComponentRenderer.php (hunk 1 above)
Comment (squrious) [reply src/TwigComponent/src/ComponentRenderer.php:184]: Good fabbot

File: src/TwigComponent/src/ComponentRenderer.php (hunk 2 above)
Comment (squrious) [reply src/TwigComponent/src/ComponentRenderer.php]: Thanks for the tip! Fixed :)

**Comment by smnandre** [comment]
Oh we lost some hair with @WebMamba last time, i will trust you on this ^^

**Review by WebMamba — APPROVED** [review]
Thanks a lot @squrious, I over thinking this part, it's much simpler and less buggy! Great job!

**Comment by kbond** [comment]
This is great, thanks Nicolas!
